src/main/java/build/buildfarm/instance/shard/ShardActionCache.java

```diff
@@ -17,47 +17,53 @@
 import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.util.concurrent.Futures.catching;
 import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static net.javacrumbs.futureconverter.java8guava.FutureConverter.toCompletableFuture;
+import static net.javacrumbs.futureconverter.java8guava.FutureConverter.toListenableFuture;
 
 import build.bazel.remote.execution.v2.ActionResult;
 import build.buildfarm.backplane.Backplane;
 import build.buildfarm.common.DigestUtil.ActionKey;
-import build.buildfarm.common.cache.CacheBuilder;
-import build.buildfarm.common.cache.CacheLoader;
-import build.buildfarm.common.cache.CacheLoader.InvalidCacheLoadException;
-import build.buildfarm.common.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.AsyncCacheLoader;
+import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.cache.CacheLoader.InvalidCacheLoadException;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.ListeningExecutorService;
 import io.grpc.Status;
 import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
 
 class ShardActionCache implements ReadThroughActionCache {
   private final Backplane backplane;
-  private final LoadingCache<ActionKey, ActionResult> actionResultCache;
+  private final AsyncLoadingCache<ActionKey, ActionResult> actionResultCache;
 
   ShardActionCache(int maxLocalCacheSize, Backplane backplane, ListeningExecutorService service) {
     this.backplane = backplane;
-    actionResultCache =
-        CacheBuilder.newBuilder()
-            .maximumSize(maxLocalCacheSize)
-            .build(
-                new CacheLoader<ActionKey, ActionResult>() {
-                  @Override
-                  public ListenableFuture<ActionResult> load(ActionKey actionKey) {
-                    return catching(
-                        service.submit(() -> backplane.getActionResult(actionKey)),
-                        IOException.class,
-                        e -> {
-                          throw Status.fromThrowable(e).asRuntimeException();
-                        },
-                        directExecutor());
-                  }
-                });
+
+    AsyncCacheLoader<ActionKey, ActionResult> loader =
+        new AsyncCacheLoader<ActionKey, ActionResult>() {
+
+          @Override
+          public CompletableFuture<ActionResult> asyncLoad(ActionKey actionKey, Executor executor) {
+            return toCompletableFuture(
+                catching(
+                    service.submit(() -> backplane.getActionResult(actionKey)),
+                    IOException.class,
+                    e -> {
+                      throw Status.fromThrowable(e).asRuntimeException();
+                    },
```

Comment:
We should investigate if this grpc-awareness is actually necessary - it really belongs at the service level to transform if used in a response. Also, this is a little cleaner as a non-throw catchingAsync.

Reply:
which part makes you say it's grpc-aware?  I'm not sure how to adjust this, as it matches to previous implementation.

Reply:
The grpc awareness is endemic to the Status translation. It's my gripe that anything below the service implementation, and above the stub, knows anything about grpc.